File: articles/biomolecules-chemical-basis-of-life/python/ligand_binding.py
"""
Ligand-binding workflow.

Run:
    python python/ligand_binding.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from biomolecule_core import ligand_fraction_bound


ARTICLE_DIR = Path(__file__).resolve().parents[1]
BINDING_PATH = ARTICLE_DIR / "data" / "ligand_binding.csv"


def main() -> None:
    """Calculate fractional occupancy for ligand-binding scenarios."""

    binding = pd.read_csv(BINDING_PATH)

    binding["fraction_bound"] = ligand_fraction_bound(
        binding["ligand_uM"].to_numpy(dtype=float),
        binding["Kd_uM"].to_numpy(dtype=float),
    )

    binding["fraction_unbound"] = 1 - binding["fraction_bound"]

    print(binding.round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/biomolecules-chemical-basis-of-life/r/ligand_binding.R
# Ligand-binding workflow in R.

binding_path <- file.path("data", "ligand_binding.csv")

if (!file.exists(binding_path)) {
  binding_path <- file.path("..", "data", "ligand_binding.csv")
}

binding <- read.csv(binding_path)

binding$fraction_bound <- binding$ligand_uM / (binding$Kd_uM + binding$ligand_uM)
binding$fraction_unbound <- 1 - binding$fraction_bound

print(round(binding, 5))

ligand <- seq(0.1, 50, length.out = 300)
Kd <- 8

binding_curve <- data.frame(
  ligand = ligand,
  fraction_bound = ligand / (Kd + ligand)
)

print(head(round(binding_curve, 5), 12))
print(tail(round(binding_curve, 5), 12))
